src/hooks/useFetch.js

```diff
@@ -4,103 +4,91 @@
-        try {
-          const response = await fetch(url, { headers, signal, ...options });
-          const { ok, status } = response;
-          if (ok) {
-            const responseData = await response.json();
-            const { error: responseError } = responseData;
-            if (responseError) throw new BugsnagError(responseError, { url, headers });
-            if (!responseData.length) throw new Error(ERROR_CODES.EMPTY_RESPONSE);
-            if (isMounted.current) setData(responseData);
-          } else {
-            switch (status) {
-              case 400:
-              default:
-                throw new Error(ERROR_CODES.CONNECTION_FAILURE);
-              case 401:
-                throw new BugsnagError(ERROR_CODES.INVALID_PASSWORD, { url, headers });
-            }
-          }
-        } catch (responseError) {
-          if (isMounted.current) {
-            if (!isValidError(responseError)) {
-              responseError.message = ERROR_CODES.CONNECTION_FAILURE;
-            }
-            setError(responseError);
-          }
-        }
-      })();
-      const timer = setTimeout(() => {
-        if (isMounted.current) setData();
-        abortController.abort();
-      }, timeout);
-      return () => {
-        isMounted.current = false;
-        clearTimeout(timer);
-        abortController.abort();
-      };
-    }
+    _isMounted.current = true;
     return () => {
-      isMounted.current = false;
+      _isMounted.current = false;
+      reset();
+    };
+  }, []);
+
+  const _fetch = async (path, init, opts) => {
+    const { responseHandler = res => res, errorHandler = err => err, timeout = 10000 } = opts;
+
+    const isMounted = () => _isMounted.current;
+    const isUnblocked = () => !_isBlocked.current;
+    const isMountedAndUnblocked = () => isMounted() && isUnblocked();
+
+    const onResponse = async res => {
+      const processedResponse = await responseHandler(res);
+      doIf(isMounted, setResponse, processedResponse);
+    };
+
+    const onError = async err => {
+      const processedError = await errorHandler(err);
+      doIf(isMounted, setError, processedError);
+    };
+
+    const beforeFetch = () => {
+      _isBlocked.current = true;
+      // eslint-disable-next-line no-undef
+      _controller.current = new AbortController();
+      _timer.current = setTimeout(() => {
+        if (_controller.current) _controller.current.abort();
+      }, timeout);
+      setIsLoading(true);
     };
-  }, [url]);
 
-  const isFetching = !(data || error);
-  return [data, error, isFetching];
+    const tryFetch = async () => fetch(url + path, { ...init, signal: _controller.current.signal });
+
+    const afterFetch = () => {
+      _isBlocked.current = false;
+      clearTimeout(_timer.current);
+      setIsLoading(false);
+    };
+
+    const doFetch = async () => {
+      try {
+        doIf(isMountedAndUnblocked, beforeFetch);
+        const res = await doIf(isMounted, tryFetch);
```

Comment:
As `beforeFetch` sets `isBlocked`, this can't be wrapped in that conditional, too! The idea would be that we shouldn't get to this point if its blocked (as `doFetch` shouldn't be executed, the extra conditional on `beforeFetch` is just added protection against race conditions!). In any case, on revisiting this it's quite clear that the `isBlocked` logic has become really convoluted, so I think it might be better to loosen some of the protections against the more improbable race conditions!